### PR TITLE
857/raster driver

### LINKED ISSUE #857: Create `raster` driver
### Kind of request

Changing existing functionality

### Enhancement Description

Create a `raster` driver using the `rasterio` package.

Also: at https://github.com/Deltares/hydromt/blob/v0.9.4/hydromt/data_adapter/rasterdataset.py#L661 zoom levels are parsed based on the rasterio metadata.

In gis/raster.py there is a to_slippy_tiles function that results in a DataCatalog entry referring to the the previous raster driver, which should refer to a new driver #875 . This behaviour can be temporarily be deprecated.

### Use case

Open data using `rasterio`.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Jaapel**
## Issue addressed
Closes #857 

## Explanation
- Created the raster driver based on functionality previously in `io.open_raster` and `io.open_mfraster`
- Added `SourceMetadata` to the `read` and `read_data` signatures of drivers. The raster driver needs this, thus `metadata` is not only for the `data_adapter`.
- Also did some pre-work for #875 while migrating old code.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `v1`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed

### CODE REVIEW

File: hydromt/drivers/dataframe/dataframe_driver.py
Comment (Jaapel) [inline hydromt/drivers/dataframe/dataframe_driver.py:24]: @savente93 I can also make the metadata optional (would just have to move the Pydantic model to circumvent cyclic imports), what would you prefer?

File: hydromt/metadata_resolver/slippy_map_tile_resolver.py
Comment (savente93) [inline hydromt/metadata_resolver/slippy_map_tile_resolver.py:121]: Sonar cloud remarks that this is duplicated from `data_adapter/rasterdataset.py` if it's no longer needed there please remove it, or if it is, could we move it to a common place like the utils?

File: hydromt/metadata.py
Comment (savente93) [inline hydromt/metadata.py:1]: Should this be here in the root or in `_type_def` ?

File: hydromt/config.py
Comment (savente93) [inline hydromt/config.py:1]: is this necessary? seems a little overkill for something we don't really offer any customization on. Would love to understand the intention behind it.

File: hydromt/data_source/__init__.py
Comment (savente93) [inline hydromt/data_source/__init__.py:4]: don't think the noqa is necessary anymore since we export it in the `__all__`  right?

File: tests/data_adapter/test_data_adapter.py
Comment (savente93) [inline tests/data_adapter/test_data_adapter.py:609]: issue? :)

File: tests/drivers/geodataset/test_vector_driver.py
Comment (savente93) [inline tests/drivers/geodataset/test_vector_driver.py:67]: this seems accidental. Is it?

**Review by savente93 — COMMENTED** [review]
I haven't really read into slippy tiles or map sctacks so I'm not totally sure what they do, but the tests look good, and I saw nothing weird so I'm good with it. THere are just a few small comments, but once those are resolved I'll be happy to approve

File: hydromt/metadata_resolver/slippy_map_tile_resolver.py
Comment (Jaapel) [reply hydromt/metadata_resolver/slippy_map_tile_resolver.py:121]: I was working on getting the slippy map tiling to work with the new architecture, but I found out while working that we decided to postpone to beta. So I already did some work, did not want to toss it.

File: hydromt/metadata.py
Comment (Jaapel) [reply hydromt/metadata.py:1]: I can move it. It should have it's own file I think

File: hydromt/config.py
Comment (Jaapel) [reply hydromt/config.py:1]: I think that certain things is HydroMT are settings. Pydantic Settings are easily settable using environment variables, so that is why I chose it. I think more settings belong here, definitively when HydroMT starts to do more (multiprocessing, secrets, etc.)

Comment (savente93) [reply hydromt/config.py:1]: That makes sense, in that case we can keep it

File: hydromt/data_source/__init__.py
Comment (Jaapel) [reply hydromt/data_source/__init__.py:4]: noqa is needed still, as the order of imports is important

File: hydromt/metadata.py
Comment (savente93) [reply hydromt/metadata.py:1]: It's not a very big deal, so I'll leave it up to you

**Review by savente93 — APPROVED** [review]
well done!

**Pull request merged**
